package ammo.protocol;

option java_package = "edu.vu.isis.ammo.coreapplication.pb";

message DataMessage {
  required string uri = 1;
  optional string mime_type = 3;
  optional bytes data = 2;
}

message PushAcknowledgement {
  required string uri = 1;
}

//Sent on initial connection to gateway to authenticate user
message AuthenticationMessage {
  required string device_id = 1;
  required string user_id = 2;
  required string user_key = 3;
}

message SubscribeMessage {
  required string mime_type = 1;
  optional string query = 2;
}

//Sent by gateway to show result of authentication
message AuthenticationResult {
  enum Status {
    FAILED = 0;
    SUCCESS = 1;
  }
  required Status result = 1;
  optional string message = 2;
}

message PullRequest {
  required string request_uid = 1;
  required string plugin_id = 2;
  required string mime_type = 3;
  required string query = 4;
  optional string projection = 5;
  optional uint32 max_results = 6 [default = 0]; //0 means infinite results--  send as many as possible
  optional uint32 start_from_count = 7 [default = 0];
  optional bool live_query = 8 [default = false];
}

message PullResponse {
  required string request_uid = 1;
  required string plugin_id = 2;
  required string mime_type = 3;
  required string uri = 4;
  required bytes data = 5;
}

message Heartbeat {
  required uint64 sequence_number = 1;
}

message MessageWrapper {
  enum MessageType {
    DATA_MESSAGE = 0;
    AUTHENTICATION_MESSAGE = 1;
    AUTHENTICATION_RESULT = 2;
    SUBSCRIBE_MESSAGE = 4;
    PUSH_ACKNOWLEDGEMENT = 5;
    PULL_REQUEST = 6;
    PULL_RESPONSE = 7;
    HEARTBEAT = 8;
  }

  required MessageType type = 1;
  optional string session_uuid = 2;
  
  optional DataMessage data_message = 3;
  optional AuthenticationMessage authentication_message= 4;
  optional AuthenticationResult authentication_result = 5;
  optional SubscribeMessage subscribe_message = 6;
  optional PushAcknowledgement push_acknowledgement = 7;
  optional PullRequest pull_request = 8;
  optional PullResponse pull_response = 9;
  optional Heartbeat heartbeat = 10;
}
